bytes2hexstring hook

RaoulSchaffranek opened the pull request:
This PR adds a new hook implementation for converting byte arrays into hex strings.
This is needed by the kontrol-node to encode memory as JSON strings efficiently.

RaoulSchaffranek committed:
WIP: bytes2hexstring

RaoulSchaffranek committed:
Run clang-format

RaoulSchaffranek committed:
Use pointer arithmetic over array index syntax

RaoulSchaffranek committed:
Fix coding style

RaoulSchaffranek committed:
Fix coding style

RaoulSchaffranek committed:
Ignore shellcheck false positive

RaoulSchaffranek committed:
Move shellcheck directive

tothtamas28 commented on runtime/strings/bytes.cpp:
```diff
@@ -162,6 +163,24 @@ SortBytes hook_BYTES_string2bytes(SortString s) {
+         '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
+  auto *result
+      = static_cast<string *>(kore_alloc_token(sizeof(string) + len * 2));
+  for (size_t i = 0; i < len; i++) {
```
I see that this pattern is repeated in `bytes.cpp`, but if `len > len(b)`, this loop will produce an out of bounds error. So consider either checking this as a precondition, or obtaining the length in the function body.

tothtamas28 commented on runtime/strings/bytes.cpp:
```diff
@@ -162,6 +163,24 @@ SortBytes hook_BYTES_string2bytes(SortString s) {
+  return result;
+}
+
+SortString hook_BYTES_bytes2hexstring(SortBytes b) {
```
Consider `bytes2hex` for a snappier name.

Also, consider adding a doc comment.

tothtamas28 commented on unittests/runtime-strings/bytestest.cpp:
```diff
@@ -171,6 +172,19 @@ BOOST_AUTO_TEST_CASE(bytes2string) {
+  BOOST_CHECK(res != empty);
+  BOOST_CHECK_EQUAL(len(res), 0);
+
+  auto _00ff = make_string("\x00\xff", 2);
```
This is not a great test input, as it will mask the error where characters for a given byte are in the wrong order in the output. Maybe `"\x01\xef"`?

tothtamas28 replied:
Also, consider adding a test case for other boundary values, in particular the empty string.

RaoulSchaffranek replied:
Good catch! I improved the test case. The empty string is already covered.

tothtamas28 commented:
@jberthold, is there a spreadsheet or similar for documenting hook implementations?

RaoulSchaffranek committed:
Address PR feedback

- Added code comment
- Renamed hook: bytes2hexstring -> bytes2hex
- Improved test case
- Use safe length to avoid potential out-of-bounds error

RaoulSchaffranek committed:
Run clang-format

RaoulSchaffranek committed:
Undo changes to bytes2string

tothtamas28 commented on runtime/strings/bytes.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <algorithm>
+#include <array> // Add this include for std::array
```
```suggestion
#include <array>
```

tothtamas28 reviewed (commented):
LGTM. I'll leave the final approval to @Robertorosmaninho.

RaoulSchaffranek committed:
Remove code comment

Co-authored-by: Tamás Tóth <tothtamas28@users.noreply.github.com>

Robertorosmaninho commented on bin/llvm-krun:
```diff
@@ -9,6 +9,7 @@ expanded_input_file="$(mktemp tmp.in.XXXXXXXXXX)"
 parser_file="$(mktemp tmp.parse.XXXXXXXXXX)"
 temp_inputs=()
 
+# shellcheck disable=SC2329
```
Sorry, why is this needed here?

RaoulSchaffranek replied:
The shell check CI workflow kept failing, see for example this action: https://github.com/runtimeverification/llvm-backend/actions/runs/17319617915/job/49169763898

If I understand correctly, it's a false positive because the function is used in the following line. So I marked it to skip the check.

It's not related to this PR, I think, but maybe some CI flakiness caused by a shell check update.

Robertorosmaninho replied:
Got it! Thanks for clarifying it!

Robertorosmaninho reviewed (approved):
LGTM, just one question:
